bin/install

```diff
@@ -1,105 +1,149 @@
 #!/usr/bin/env bash
 
+# We have no need for filename expansion / globbing,
+# so we'll disable it here.
+# We'll take advantage of the safety this confers when doing something like:
+#
+#   confg_args=(--some arg $USER_EXTRA_ARGS)
+#
+# Without set -f, $USER_EXTRA_ARGS can be subject to any glob patterns therein!
+set -f
+
+# fail-fast
+set -euo pipefail
+
 install_luaJIT() {
   local install_type=$1
   local raw_version=$2
   local install_path=$3
-  local luaJIT_version=$(echo $raw_version | sed -e 's/--.*$//g')
-  local lua_rocks_version=$(echo $raw_version | sed -e "s/^.*--//g")
+  local luaJIT_version="${raw_version%%--*}"
+  local lua_rocks_version="${raw_version##*--}"
 
-  if [ "$TMPDIR" = "" ]; then
-    local tmp_download_dir=$(mktemp -d)
+  local tmp_download_dir
+  if [[ "${TMPDIR-}" == "" ]]; then
+    tmp_download_dir=$(mktemp -d)
   else
-    local tmp_download_dir=$TMPDIR
+    tmp_download_dir=$TMPDIR
   fi
 
-  local luaJIT_source_path=$(get_luaJIT_download_file_path $install_type $luaJIT_version $tmp_download_dir)
-  local lua_rocks_source_path=$(get_lua_rocks_download_file_path $install_type $lua_rocks_version $tmp_download_dir)
-  download_luaJIT_source $install_type $luaJIT_version $luaJIT_source_path
-  download_lua_rocks_source $install_type $lua_rocks_version $lua_rocks_source_path
+  # canonicalize the path.
+  # this has several benefits:
+  #  1. if the luajit/luarocks build-systems also does this (hint: they do),
+  #     then the path we use here will be guaranteed to be identical.
+  #  2. this removes silly stuff like //, /./ throughout the path,
+  #     and removes a trailing slash (e.g. macOS's $TMPDIR).
+  tmp_download_dir=$(realpath -- "$tmp_download_dir")
+
+  local luaJIT_source_path
+  luaJIT_source_path=$(get_luaJIT_download_file_path "$install_type" "$luaJIT_version" "$tmp_download_dir")
+  local lua_rocks_source_path
+  lua_rocks_source_path=$(get_lua_rocks_download_file_path "$install_type" "$lua_rocks_version" "$tmp_download_dir")
+  download_luaJIT_source "$install_type" "$luaJIT_version" "$luaJIT_source_path"
+  download_lua_rocks_source "$install_type" "$lua_rocks_version" "$lua_rocks_source_path"
 
   # running this in a subshell
   # because we don't want to disturb current working dir
   (
-    cd ${luaJIT_source_path}
+    local line
+
+    cd "${luaJIT_source_path}" || exit 1
+
+    local luaJIT_configure_options=()
+    while IFS='' read -r line; do
+      luaJIT_configure_options+=("$line")
+    done < <(construct_luaJIT_configure_options)
```

Comment:
So... Ideally we'd model "multiple things" (like configuration parameters) as arrays of strings. And `bash` does have arrays. But what it doesn't have is any way to _return_ an array. You can abuse globals, or you can use [namerefs](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameters.html) if you can count on a suitably non-ancient version of `bash` (but we can't do that because macOS is stuck on 3.2.57, copyright 2007 😢). You could pass around strings separated by spaces, and let `bash` do it's thing where it splits things up based on `$IFS`, but now you can't tell if "/home/charles strahan" is two elements (`/home/charles`, /`strahan`) or one path that happens to have a space in it.

As a compromise, we'll return arrays as elements separated by newlines, piped to stdout, and on the receiving side we'll iterate over each line of stdin and build up an array.

So the pattern will be something like:

```bash
# print every element of the array on a separate line
printf "%s\n" "${some_array_here[@]}"
```

and

```bash
local luaJIT_configure_options=()
# read the output of `returns_some_array` line-by-line and construct an array.
# IFS='' prevents splitting on any other character than "\n",
# and the -r prevents `read` from trying to be helpful by
# converting a "\\n" -- i.e. literally a slash followed by a 'n verbatim -- into a '\n' character)
while IFS='' read -r line; do
  luaJIT_configure_options+=("$line")
done < <(returns_some_array)
```

Reply:
I'd love to rewrite

```bash
local luaJIT_configure_options=()
while IFS='' read -r line; do
  luaJIT_configure_options+=("$line")
done < <(construct_luaJIT_configure_options)
```

as

```bash
mapfile -t luaJIT_configure_options < <(construct_luaJIT_configure_options)
```

but that's not an option for macOS.

Details: https://www.shellcheck.net/wiki/SC2207

Reply:
I'd have never caught this.